Extend the options for coverage instrumentation

forejtv opened the pull request:
This PR adds a `cover-only` option that mirrors what `cover-function-only` is doing, but allows the user to specify that a single file can be instrumented, rather than just a single function. As a side-effect of the changes, I fixed a bug in `cover-function-only`: before, we were matching function name as substring, so for the main function `Foo.foo` we'd also have instrumented `Foo.foolish` when `cover-function-only` was used, which was incorrect.

~There are no tests with this PR because the command-line option `cover-only` or `cover-function-only` is not enabled in any of the opensource binaries.~

Commit:
Refactor function filter for coverage instrumentation.

This will make it possible to create filters that depend on more
elaborate properties of functions, as opposed to the current behaviour
when we could basically just filter using function's name.

kroening commented:
I would certainly not add the pointer to the symbol table in `config`.

thk123 commented on jbmc/src/java_bytecode/java_entry_point.cpp:
```diff
@@ -537,6 +537,7 @@ main_function_resultt get_main_symbol(
       symbol != nullptr,
       "resolve_friendly_method_name should return a symbol-table identifier");
 
+    config.main_symbol = symbol;
```
:question: The less we use config, the better, is this really the best way to achieve what you want?

thk123 replied:
Though I see was already using it: https://github.com/diffblue/cbmc/pull/4180/commits/85348809d2b52d3230140511be42e45f3ae28732#diff-5c5e737bf6dd6d1aa2892f71af61a580L175

thk123 replied:
Out of band discussion: Vojta will add some kind of utility to start with that allows deducing the entry point from a model (possibly the model show _know_ the entry point function?

thk123 commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -43,33 +44,33 @@ bool internal_functions_filtert::operator()(
   const goto_functionst::goto_functiont &goto_function) const
 {
-  (void)function_id; // unused parameter
+  (void)function; // unused parameter
```
:pick: If unused just delete the name of the field.

peterschrammel replied:
Note that you can't refer to it from doxygen then...

thk123 replied:
Why would you want to document an unused parameter?

hannes-steffenhagen-diffblue replied:
@thk123 the way we’ve set up linting we require all parameters to a function to be documented if any are

thk123 commented on src/goto-instrument/cover_filter.h:
```diff
@@ -85,7 +86,7 @@ class function_filterst
   /// \param goto_function: goto function
   bool operator()(
-    const irep_idt &identifier,
+    const symbolt &identifier,
```
:pick: This is missing `override`

forejtv replied:
Are you sure? The compiler does not like when I put `override` to this function.

hannes-steffenhagen-diffblue replied:
This doesn’t override anything, therefore override can’t and shouldn’t be used

thk123 replied:
My mistake

thk123 commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -44,6 +44,20 @@ bool internal_functions_filtert::operator()(
+  const symbolt &function,
+  const goto_functionst::goto_functiont &goto_function) const
+{
+  (void)goto_function; // unused parameter
```
:pick: Just don't give the unused param a name

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -169,7 +169,16 @@ void parse_cover_options(const cmdlinet &cmdline, optionst &options)
+  std::string cover_only = cmdline.get_value("cover-only");
+
+  if(!cover_only.empty() && cmdline.isset("cover-function-only"))
+    throw "at most one of --cover-only and --cover-function-only can be used";
```
Throwing naked strings makes life hard, probably want `invalid_command_line_argument_exceptiont`

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -225,13 +234,26 @@ std::unique_ptr<cover_configt> get_cover_config(
+  {
+    msg.error() << "Argument to --cover-only not recognised: "
+                << cover_only << messaget::eom;
+    return {};
```
inconsistent error handling - should work the same as providing two conflicting cover commands

forejtv replied:
Are you sure you want me to change this? What I've done there is I've used the same syntax as other error messages in the same function.

thk123 replied:
Yeah OK you're probably right - looked weird in the context of the place where you were throwing an exception.

thk123 reviewed (commented):
I see the tests in TG, but they probably want to live in CBMC since that's where the code is
Might be faster & more precise to write them as unit tests, but not a requirement.

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -315,9 +341,11 @@ void instrument_cover_goals(
   message_handlert &message_handler)
 {
+  const symbolt *function_symbol =
+    function.get_symbol_table().lookup(function.get_function_id());
```
Use `lookup_ref` to avoid storing in a pointer that you immediately dereference.

peterschrammel commented on src/goto-instrument/cover.cpp:
```diff
@@ -231,7 +234,27 @@ std::unique_ptr<cover_configt> get_cover_config(
+  }
+  else if(!cover_only.empty())
+  {
+    msg.error() << "Argument to --cover-only not recognised: "
```
recognized

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 016cea05e36bd934bf85a84e63630fb6793ca590).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100922275

Commit:
Unit tests for cover-only

forejtv commented:
Thanks for all the comments. This should be ready to ready to re-review.

Note that I have not done anything about the unused parameters. Unless I misunderstand the comments, @thk123 and @peterschrammel don't agree which approach is better, so I went for a status quo and kept the approach that was in there before.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 843de11cabf97babf5c194d584d11681bef02364).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100955487
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aecce9c36ca80216544ea29aef71cdaf6e4aadcc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100958306

thk123 commented on unit/cover_only.cpp:
```diff
@@ -0,0 +1,142 @@
+  GIVEN("cover-only set to an unrecognised value")
+  {
+    optionst options;
+    options.set_option("cover-only", "diplodocus-with-a-duvet");
```
:green_heart:

thk123 commented on unit/Makefile:
```diff
@@ -16,6 +16,7 @@ SRC += analyses/ai/ai.cpp \
        analyses/does_remove_const/is_type_at_least_as_const_as.cpp \
        big-int/big-int.cpp \
        compound_block_locations.cpp \
+       cover_only.cpp \
```
:no_entry_sign: Please put the test inside `goto-instrument/cover/cover_only.cpp`

thk123 commented on unit/cover_only.cpp:
```diff
@@ -0,0 +1,142 @@
+
+Module: Tests for coverage instrumentation config object and filters
+
+Author: Vojtech Forejt <v48@diffblue.com>, 2018
```
:pick: 2019

Also in general Diffblue contributors tend to put Diffblue Ltd as the author I think?

thk123 commented on unit/cover_only.cpp:
```diff
@@ -0,0 +1,142 @@
+      THEN("A config is obtained that matches all the symbols")
+      {
+        REQUIRE(ret);
+        REQUIRE(ret);
```
:pick: Delete the second (duplicate) `REQUIRE(ret)`

thk123 commented on unit/cover_only.cpp:
```diff
@@ -0,0 +1,142 @@
+      THEN("A config is obtained that matches all the symbols")
+      {
+        REQUIRE(ret);
+        REQUIRE(ret);
```
:pick: Delete the second (duplicate) `REQUIRE(ret)`

thk123 commented on unit/cover_only.cpp:
```diff
@@ -0,0 +1,142 @@
+  symbol_tablet symbol_table;
+  symbolt symbol_Afoo = create_new_symbol("java::A.foo", "A.java");
+
+  symbolt symbol_Afoolish = create_new_symbol("java::A.foolish", "A.java");
```
Is it worth testing a couple of other things:
```
"java::A$Inner.foo", "A.java"
"java::AnotherPackage.A.bar", "AnotherPackage/A.java"
```

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -231,7 +237,29 @@ std::unique_ptr<cover_configt> get_cover_config(
 
+  std::string cover_only = options.get_option("cover-only");
+
+  DATA_INVARIANT(
```
This seems like an unusual approach since at run time you can't really pick whether you pass an argument in. Is there a reason to not just make the main symbol required? Or perhaps refactoring into two separate functions, one for `cover-only` configuration and one for the other stuff, so it is clear at compile time when a main symbol is required.

forejtv replied:
Are you talking about the data invariant? If yes, then at runtime you do not pick whether you pass the argument in, this is up to the caller(s) who have this hardcoded.

The reason I did this was that coverage instrumentation (`instrument_cover_goals`) is used by multiple callers (it's used in multiple binaries), and if I wanted to always pass main symbol in, I'd have to modify all callers to find what the main symbol is, which is non-trivial. And in the end only one of the callers happens to pass `cover-only` which uses the main symbol.

I can go with the two-function approach I suppose, for this to avoid code duplication it will mean that `instrument_cover_goals` will not call `get_cover_config`, but instead it will take config as an argument and it will be the caller's responsibility to call the right variant of `get_cover_config` functions (there will be two) and pass its result on. The issue is that this will change the arguments of `instrument_cover_goals`, and I'm not sure if that's ok to do: given that one of the overloads is marked as `DEPRECATED`, I suppose that there is an expectation to keep the arguments stable-ish.

thk123 replied:
Your proposed solution sounds better, if I understand it correctly as it separates out the concern of dealing with `cover-only` from a function that otherwise doesn't have to worry about it. I'm not sure I follow your concern with your solution - you are already changing the arguments - but this isn't a blocking comment if you think it is too much.

thk123 commented on src/goto-instrument/cover.h:
```diff
@@ -66,6 +70,13 @@ void instrument_cover_goals(
 
 void parse_cover_options(const cmdlinet &, optionst &);
 
+bool instrument_cover_goals(
```
This one definitely shouldn't taken an optional, since there is already an overload that doesn't require one at all - how would I pick between calling with with `{}` and without the argument entirely?

forejtv replied:
The reason I left the other one there was backwards compatibility of public methods. I'm happy to go with either solution, I can change all known usages of the not-even-optional overload to using optional. Up to you @thk123 and whoever else reviews this PR.

thk123 replied:
Optional is not required for backwards compatibility, as there already exists an overload that doesn't take the parameter. Happy to talk through this with you, but this one should really be addressed IMO.

thk123 reviewed (commented):
Looks much better, the unit test isn't in the [right place](https://github.com/diffblue/cbmc/blob/develop/CODING_STANDARD.md#unit-tests) 

I also have some questions about the how the main symbol is passed around, but maybe there are good practical reasons why it is this way.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7dfbf3beae63ebe39ae15194966385152c816f48).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101296295

forejtv commented:
@thk123 I have put the config for coverage out of the instrumentation itself, so this is now ready to re-review.

thk123 commented on jbmc/src/jdiff/jdiff_parse_options.cpp:
```diff
@@ -321,7 +321,12 @@ bool jdiff_parse_optionst::process_goto_program(
     {
-      if(instrument_cover_goals(options, goto_model, get_message_handler()))
+      const auto cover_config = get_cover_config(
+        options, goto_model.symbol_table, log.get_message_handler());
```
:question: Why do the two calls use different message handlers?

forejtv replied:
copy-paste mistake, well spotted!

thk123 reviewed (approved):
Looks much better IMO.

tautschnig commented on src/goto-instrument/cover.cpp:
```diff
@@ -184,11 +176,13 @@ void parse_cover_options(const cmdlinet &cmdline, optionst &options)
 /// \return a cover_configt on success, or null otherwise.
 std::unique_ptr<cover_configt> get_cover_config(
   const optionst &options,
+  const optionalt<symbolt> &main_symbol,
```
Is this actually used?

forejtv replied:
Yes it is eventually, I did a bad rebase and it was not used in the particular commit.

kroening commented:
Do you want to make the coverage criterion part of `cover_configt`?

kroening commented on src/goto-instrument/cover.h:
```diff
@@ -46,18 +46,26 @@ struct cover_configt
-  const optionst &, const symbol_tablet &, message_handlert &);
+  const optionst &,
+  const symbolt &,
+  const symbol_tablet &,
```
I'd add a declarator to the `symbolt` parameter, to tell me what it does. All the others are obvious.

kroening replied:
Also, given that you are passing in the symbol_table, does it not make sense to give the identifier instead?

kroening commented on src/goto-instrument/cover.cpp:
```diff
@@ -251,14 +256,59 @@ std::unique_ptr<cover_configt> get_cover_config(
+/// \param message_handler: used to log incorrect option specifications
+/// \return a cover_configt on success, or null otherwise.
+std::unique_ptr<cover_configt> get_cover_config(
+  const optionst &options,
```
We are doing C++11 -- so no need for the pointer; simply return the thing, and make sure you've got a working rvalue constructor.

kroening replied:
Ah, and if you also want to signal error, use `optionalt`, not a null pointer.

hannes-steffenhagen-diffblue replied:
IMHO if a `configt` can't be constructed because there's an error with the options it should just throw an appropriate exception, like `invalid_command_line_optiont`.

kroening commented on src/goto-instrument/cover.cpp:
```diff
@@ -251,14 +256,59 @@ std::unique_ptr<cover_configt> get_cover_config(
+
+  std::string cover_only = options.get_option("cover-only");
+  const symbolt main_symbol = symbol_table.lookup_ref(main_id);
+
```
only do the lookup if needed, i.e., in the "function" case. Also, use a reference, don't copy the symbol.

forejtv replied:
It is used in two if branches (function and file), but ok, I'll put the lookup in each.

kroening commented on src/goto-instrument/cover.cpp:
```diff
@@ -251,14 +256,59 @@ std::unique_ptr<cover_configt> get_cover_config(
+std::unique_ptr<cover_configt> get_cover_config(
+  const optionst &options,
+  const irep_idt &main_id,
+  const symbol_tablet &symbol_table,
```
Is the name "main_id" appropriate? I understand that this is the "function" for "cover this function".

forejtv replied:
renamed to main_function_id

kroening commented on src/goto-instrument/cover.h:
```diff
@@ -68,12 +75,14 @@ void parse_cover_options(const cmdlinet &, optionst &);
 bool instrument_cover_goals(
   const optionst &,
+  const cover_configt &,
   const symbol_tablet &,
```
Does this really need both the `optionst` and the `cover_configt`?

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover.cpp:
```diff
@@ -176,6 +176,7 @@ void parse_cover_options(const cmdlinet &cmdline, optionst &options)
   }
   options.set_option("cover-include-pattern", cover_include_pattern);
   options.set_option("no-trivial-tests", cmdline.isset("no-trivial-tests"));
+  options.set_option("cover-function-only", cmdline.isset("cover-function-only"));
```
The commit message says something about refactoring, but this adds a new entry to `options`? Was this meant to be in one of the later commits?

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover.cpp:
```diff
@@ -252,6 +256,51 @@ std::unique_ptr<cover_configt> get_cover_config(
+  }
+  else if(!cover_only.empty())
+  {
+    msg.error() << "Argument to --cover-only not recognized: " << cover_only
```
`throw invalid_command_line_argument_exceptiont{this message, "--cover-only"}`

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover.h:
```diff
@@ -46,18 +45,26 @@ struct cover_configt
   message_handlert &message_handler);
 
+std::unique_ptr<cover_configt>
+get_cover_config(const optionst &, const symbol_tablet &, message_handlert &);
```
Should be documented, as should the overload with more parameters.

Actually, the different overloads seem to be kind of doing different things, so maybe they should have different names? One seems to create a basic `cover_configt` the other just adds a bunch of filters? Maybe it should just take the `cover_config` by ref and modify it?

forejtv replied:
This is documented in the cpp file, it was the way this was done before I edited the files.

I'd keep the same names, the overload with extra parameters does strictly more than the overload without it, and I believe this is a standard characteristics of overloads.

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover_filter.cpp:
```diff
@@ -43,6 +43,34 @@ bool internal_functions_filtert::operator()(
+  const symbolt &function,
+  const goto_functionst::goto_functiont &goto_function) const
+{
+  (void)goto_function; // unused parameter
```
Put the documentation to the declaration, then you don't need this at the implementation because you can leave the parameter name off

forejtv replied:
I'd prefer to keep this as it it, since I was just re-using the style the file had before my changes.

hannes-steffenhagen-diffblue commented on unit/goto-instrument/cover/cover_only.cpp:
```diff
@@ -0,0 +1,157 @@
+
+Module: Tests for coverage instrumentation config object and filters
+
+Author: Diffblue Ltd, 2019
```
2019 is not an author. We have a `Date: ` field in some versions of this header, but I'm not sure how useful that even is considering this is presumably going to be changed many times and we likely won't update the date field every time?

hannes-steffenhagen-diffblue commented on unit/goto-instrument/cover/cover_only.cpp:
```diff
@@ -0,0 +1,157 @@
+#include <util/symbol_table.h>
+#include <util/ui_message.h>
+
+symbolt create_new_symbol(const irep_idt &name, const irep_idt &file_name)
```
:pick: This is the sort of name we could easily use in another place over time, I'd wrap this up in an anonymous namespace to prevent linking errors in the future

hannes-steffenhagen-diffblue commented:
In the second commit message s/Tkae/Take

Commit:
get_cover_config is not a pointer and proper exception is thrown if it can't be constructed

Commit:
put main_symbol into the if branches and make it a reference

Commit:
rename main_id to main_function_id

Commit:
instrument_cover_goals does not need optionst

Commit:
get_cover_config is now called separately from instrument_cover_goals

Commit:
Add cover-only and replace pattern handling of main function.

This commit adds a new command-line option in the code (not on the
command line itself), `cover-only`, which takes either `function` or
`file` as parameter. The former is equivalent to `cover-function-only`,
the later instruments the file which contains the main function.

As part of the changes, instead of matching the function name using
regex for coverage instrumentation we use the actual function symbol.

forejtv commented:
@kroening @hannes-steffenhagen-diffblue This is ready for a re-review.

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover.cpp:
```diff
@@ -205,33 +210,25 @@ std::unique_ptr<cover_configt> get_cover_config(
-    msg.error() << "assertion coverage cannot currently be used together with "
-                << "other coverage criteria" << messaget::eom;
-    return {};
+    std::stringstream s;
```
⛏ this could be an ostringstream

hannes-steffenhagen-diffblue reviewed (approved):
I’d still prefer if the documentation would be moved to the header but I guess that doesn’t need to happen right now

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 83b336292e01f9fb57493fcc43d20a132f17fe75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102413217
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

Commit:
Fix goal filters bug

Goal filters are created as part of cover_config, and then passed by
reference to instrumenters. When cover_config gets assigned, we end up
with pointers pointing to nowhere. The fix is to make sure that goal
filters are created on a heap.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d915a096a0d7e775e05cc295f1ea7b44c8a71410).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102511210
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

smowton reviewed (approved):
Haven't reviewed, just freeing @forejtv from code-owner hell
